bumble/utils.py

```diff
@@ -184,38 +229,48 @@ def close(self) -> None:
 _T = TypeVar('_T')
 
 
-class AbortableEventEmitter(EventEmitter):
-    def abort_on(self, event: str, awaitable: Awaitable[_T]) -> Awaitable[_T]:
-        """
-        Set a coroutine or future to abort when an event occur.
-        """
-        future = asyncio.ensure_future(awaitable)
+def cancel_on_event(
+    emitter: pyee.EventEmitter, event: str, awaitable: Awaitable[_T]
+) -> Awaitable[_T]:
+    """Set a coroutine or future to cancel when an event occur."""
+    future = asyncio.ensure_future(awaitable)
+    if future.done():
+        return future
+
+    def on_event(*args, **kwargs) -> None:
+        del args, kwargs
         if future.done():
-            return future
-
-        def on_event(*_):
-            if future.done():
-                return
-            msg = f'abort: {event} event occurred.'
-            if isinstance(future, asyncio.Task):
-                # python < 3.9 does not support passing a message on `Task.cancel`
-                if sys.version_info < (3, 9, 0):
-                    future.cancel()
-                else:
-                    future.cancel(msg)
+            return
+        msg = f'abort: {event} event occurred.'
+        if isinstance(future, asyncio.Task):
+            # python < 3.9 does not support passing a message on `Task.cancel`
+            if sys.version_info < (3, 9, 0):
+                future.cancel()
             else:
-                future.set_exception(asyncio.CancelledError(msg))
+                future.cancel(msg)
+        else:
+            future.set_exception(asyncio.CancelledError(msg))
 
-        def on_done(_):
-            self.remove_listener(event, on_event)
+    def on_done(_):
+        emitter.remove_listener(event, on_event)
 
-        self.on(event, on_event)
-        future.add_done_callback(on_done)
-        return future
+    emitter.on(event, on_event)
+    future.add_done_callback(on_done)
+    return future
+
+
+# -----------------------------------------------------------------------------
+class BaseBumbleEventEmitter(pyee.asyncio.AsyncIOEventEmitter):
```

Comment:
Maybe we could keep this name short, just calling it `EventEmitter`? Since it is in the `utils` module, that already provides a form of namespace.

Reply:
Done